### PR TITLE
client/core: Recognize missing trades.

### LINKED ISSUE #3022: core: redeem unknown match ID does not resolve.
Found by an mm user. If you have a match that makes it to redemption that is revoked by the server but you miss the revoke message, bisonw will continue to try to send a redemption but be met with the error:
```
2024-10-15 08:07:23.583 [ERR] CORE: notify: |ERROR| (order) Redeem reporting error - Error notifying DEX of redemption for match 72bdf8fc34dde82abdd1ccd8e33b2384e4282631de92e5af4089240ac4a8fc69: error sending 'redeem' message: rpc error: error code 6: unknown match ID - Order: 201196e096b15ab647bde6c0ee1328255c4939eb636249b5f1e24a2c55e0db87
2024-10-15 08:07:25.840 [DBG] CORE: Notifying DEX 127.0.0.1:17273 of our dcr swap redemption d4c7599c1d56ab34db4ff8f64392ebe0c2b49f820a39310b01168d5274433bcc:0 for match 72bdf8fc34dde82abdd1ccd8e33b2384e4282631de92e5af4089240ac4a8fc69
2024-10-15 08:07:25.840 [DBG] CORE: Soon to expire bond found: 0d1b7c9b6f133b3e52b8377e15e5c6b25eb8d0e9e971cc6ce868d9837a6b0cac:0 (dcr)
2024-10-15 08:07:25.840 [DBG] CORE: Expired bond 7ddcb0f21c59e74b3dbef09c93e472eebcb64080926b2dc03635cc65aab3de80:0 (dcr) with lock time 2024-10-15 17:05:14 +0900 JST not yet refundable according to wallet.
2024-10-15 08:07:25.855 [ERR] CORE: notify: |ERROR| (order) Redeem reporting error - Error notifying DEX of redemption for match 72bdf8fc34dde82abdd1ccd8e33b2384e4282631de92e5af4089240ac4a8fc69: error sending 'redeem' message: rpc error: error code 6: unknown match ID - Order: 201196e096b15ab647bde6c0ee1328255c4939eb636249b5f1e24a2c55e0db87
2024-10-15 08:07:28.331 [TRC] WEB[WS]: message of type 1 received for route acknotes
2024-10-15 08:07:33.339 [DBG] CORE: Notifying DEX 127.0.0.1:17273 of our dcr swap redemption d4c7599c1d56ab34db4ff8f64392ebe0c2b49f820a39310b01168d5274433bcc:0 for match 72bdf8fc34dde82abdd1ccd8e33b2384e4282631de92e5af4089240ac4a8fc69
2024-10-15 08:07:33.342 [ERR] CORE: notify: |ERROR| (order) Redeem reporting error - Error notifying DEX of redemption for match 72bdf8fc34dde82abdd1ccd8e33b2384e4282631de92e5af4089240ac4a8fc69: error sending 'redeem' message: rpc error: error code 6: unknown match ID - Order: 201196e096b15ab647bde6c0ee1328255c4939eb636249b5f1e24a2c55e0db87
```

The client never stops trying to send the redemption until restart and `dc.compareServerMatches` happens.

### PR DESCRIPTION
**Pull request by JoeGruffins**
When sending inits and redemptions, if the server does not know about the
match, assume revoked and stop sending.

closes #3022

### CODE REVIEW

**Review by martonp — APPROVED** [review]
LGTM

File: client/core/trade.go [hunk 1]
```diff
@@ -2879,6 +2879,24 @@ func (c *Core) sendRedeemAsync(t *trackedTrade, match *matchTracker, coinID, sec
 					},
 				})
 			}
+			if errors.As(err, &msgErr) && msgErr.Code == msgjson.RPCUnknownMatch {
```
Comment (buck54321) [inline client/core/trade.go:2882]: Looks like the same thing could happen when we send `msgjson.InitRoute` in `sendInitAsync`. Do we need to self-revoke there too?

**Comment by dev-warrior777** [comment]
Tested the updated code by simulating what the mm user reported i.e. He was Taker and the _assumption_ is counter-party went offline at some stage during the swap but the expected revoke did not happen for taker. 

 1. Maker goes offline after Taker contract confirmed.        (walletpair/maker killed)
     Refund issued

image


2. Maker goes offline after just Maker contract confirmed. (walletpair/maker killed)
    Refund issued .. 

image

**Comment by JoeGruffins** [comment]
@dev-warrior777 I think the error only happens when one, or both, party misses the revoke notification from server. So they continue to redeem, or already redeemed, and then try to inform the server of the redeem that server has forgotten about. So you also need to make a client miss the revoke notification to hit the path.

**Comment by dev-warrior777** [comment]
> @dev-warrior777 I think the error only happens when one, or both, party misses the revoke notification from server. So they continue to redeem, or already redeemed, and then try to inform the server of the redeem that server has forgotten about. So you also need to make a client miss the revoke notification to hit the path.

Ok .. will add your "not really revoking" returns from issue #[3022](https://github.com/decred/dcrdex/issues/3022)

File: client/core/trade.go (hunk 1 above)
Comment (JoeGruffins) [reply client/core/trade.go:2882]: I was able to cause the same error with init, so yes.

**Comment by JoeGruffins** [comment]
Also checking the error for init sends https://github.com/decred/dcrdex/compare/fdc924abe1afb4e6cba11028bf5da6d7f55fe3ef..5c34849a7f72cb3b4d7586be2f86dd79c9d6165d

Unsure if we need to check balances immediately. I assume that happens in a tick?

**Comment by dev-warrior777** [comment]
Tested redemptions again with updated code `5c34849`

[missed_redemption.zip](https://github.com/user-attachments/files/17400138/missed_redemption.zip)

File: client/core/trade.go [hunk 2]
```diff
@@ -2609,6 +2609,21 @@ func (c *Core) sendInitAsync(t *trackedTrade, match *matchTracker, coinID, contr
 					},
 				})
 			}
+			if msgErr.Code == msgjson.RPCUnknownMatch {
```
Comment (JoeGruffins) [inline client/core/trade.go:2612]: Ah, maybe error is still nil if the above errors.As failed

Comment (dev-warrior777) [reply client/core/trade.go:2612]: > Ah, maybe error is still nil if the above errors.As failed

I had nil one time while testing .. not sure why. and only for init. Does it work if you add back the As test?

Comment (JoeGruffins) [reply client/core/trade.go:2612]: Yeah should.

**Comment by JoeGruffins** [comment]
Moved everything inside the errors.As https://github.com/decred/dcrdex/compare/5c34849a7f72cb3b4d7586be2f86dd79c9d6165d..b449fa9a1f47ea41274a9a27996254fa046fdfb3

**Review by dev-warrior777 — APPROVED** [review]
Looks good with these errors.As changes

File: client/core/trade.go [hunk 3]
```diff
@@ -2600,14 +2600,30 @@ func (c *Core) sendInitAsync(t *trackedTrade, match *matchTracker, coinID, contr
...
+							matches: []*matchTracker{match},
+						},
+					})
+				} else if msgErr.Code == msgjson.RPCUnknownMatch {
+					oid := t.ID()
+					c.log.Warnf("DEX %s did not report active match %s on order %s - assuming revoked, status %v.",
+						t.dc.acct.host, match, oid, match.Status)
+					// We must have missed the revoke notification. Flag to allow recovery
+					// and subsequent retirement of the match and parent trade.
+					match.MetaData.Proof.SelfRevoked = true
```
Comment (buck54321) [inline client/core/trade.go:2618]: Don't you need to lock the `trackedTrade.mtx` to modify the `Proof`?

Comment (JoeGruffins) [reply client/core/trade.go:2619]: yes https://github.com/decred/dcrdex/compare/eef6d19a4c715c7ef42c2313756879851f0d40e2..e6533b691c41d3939fb293c3d1572154e1679b20

Comment (JoeGruffins) [reply client/core/trade.go:2619]: Did we need to hold it to read `match.Status`?

Comment (JoeGruffins) [reply client/core/trade.go:2619]: Moved it up https://github.com/decred/dcrdex/compare/e6533b691c41d3939fb293c3d1572154e1679b20..377603af2b79e07c5347ce5b3da752d3b126e5e9

**Pull request merged**